include/deal.II/base/quadrature.h

```diff
@@ -241,18 +241,116 @@ class Quadrature : public Subscriptor
   void
   serialize(Archive &ar, const unsigned int version);
 
+protected:
   /**
-   * This function returns true if the quadrature object is a tensor product
-   * of one-dimensional formulas and the quadrature points are sorted
-   * lexicographically.
+   * List of quadrature points. To be filled by the constructors of derived
+   * classes.
    */
-  bool
-  is_tensor_product() const;
+  std::vector<Point<dim>> quadrature_points;
+
+  /**
+   * List of weights of the quadrature points.  To be filled by the
+   * constructors of derived classes.
+   */
+  std::vector<double> weights;
+};
+
+
+/**
+ * This function returns true if the incoming Quadrature<dim> is of type
+ * TensorProductQuadrature<dim>.
+ */
+template <int dim>
+bool
+is_tensor_product(const Quadrature<dim> &quadrature);
+
+
+/**
+ * This class represents a Quadrature<dim> which is the tensor product of a
+ * dim 1-dimensional quadratures. The main feature of this class is that it is
+ * possible to extract the base quadratures using the get_tensor_basis()
+ * function. Being able to access these base quadratures allows for
+ * optimizations which otherwise wouldn't be possible.
+ */
+template <int dim>
+class TensorProductQuadrature : public Quadrature<dim>
+{
+public:
+  /**
+   * Alias for a quadrature that acts on an object of one dimension
+   * less. Note that a SubTensorProductQuadrature is an SubQuadrature.
+   *
+   * Since a 1-dimensional quadrature can be considered a tensor product we
+   * let the SubTensorProductQuadrature for dim=2 to be an ordinary quadrature.
+   */
+  using SubTensorProductQuadrature =
+    typename std::conditional<(1 < dim),
+                              TensorProductQuadrature<dim - 1>,
+                              Quadrature<0>>::type;
```

Comment:
Could possibly let
Quadrature<2>::SubTensorProductQuadrature be Quadrature<1> ?
Not really sure what is the best.